### PR TITLE
Return enums from model getters and make enums follow the standard java naming scheme.

### PR DESCRIPTION
**Pull request by millems**
Where appropriate, enums are now returned from model getters instead of
strings, which have been moved to a separate method that indicates its
value is a string.

Generated enums now follow the SCREAMING_CASE convention instead of the
PascalCase that was previously used.

### CODE REVIEW

File: codegen/src/test/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategyTest.java [hunk 1]
```diff
@@ -16,25 +16,80 @@
...
+        validateConversion("SSE_S3", "SSE_S3");
+        validateConversion("http1.1", "HTTP1_1");
+        validateConversion("T100", "T100");
+        validateConversion("s3:ObjectCreated:*", "S3_OBJECT_CREATED");
+        validateConversion("s3:ObjectCreated:Put", "S3_OBJECT_CREATED_PUT");
+        validateConversion("TLSv1", "TLS_V1");
+        validateConversion("TLSv1.2", "TLS_V1_2");
+        validateConversion("us-east-1", "US_EAST_1");
+        validateConversion("io1", "IO1");
+        validateConversion("application/vnd.amazonaws.card.generic", "APPLICATION_VND_AMAZONAWS_CARD_GENERIC");
```
Comment (spfink) [inline codegen/src/test/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategyTest.java]: Duplicated from line 49

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/common/test-enum-class.java [hunk 2]
```diff
@@ -37,21 +22,20 @@ private TestEnumClass(String value) {
...
+     * Use this in place of valueOf to convert the raw string returned by the service into the enum value.
      *
      * @param value
      *        real value
      * @return TestEnumClass corresponding to the value
      */
     public static TestEnumClass fromValue(String value) {
-        if (isBlank(value)) {
-            throw new IllegalArgumentException("Value cannot be null or empty!");
+        if (value == null) {
```
Comment (spfink) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/common/test-enum-class.java]: Why the change in behavior here?

Comment (millems) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/common/test-enum-class.java]: It was easier to use this way. If the service returned null, you'd have to do a null and isEmpty check before converting it to an enum. This way, a null response from the service is a null enum, and an empty-string response from the service is an UNKNOWN enum.

File: codegen/src/test/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategyTest.java (hunk 1 above)
Comment (millems) [reply codegen/src/test/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategyTest.java]: Good catch!

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java [hunk 3]
```diff
@@ -307,11 +448,29 @@ public SubTypeOne polymorphicTypeWithoutSubTypes() {
 
     /**
      * Returns the value of the EnumType property for this object.
+     * <p>
+     * If the service returns an enum value that is not available in the current SDK version, {@link #enumType} will
+     * return {@link EnumType#UNKNOWN}. The raw value returned by the service is available from {@link #enumTypeString}.
```
Comment (shorea) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java]: We should use SDK_UNKNOWN or something like that to prevent clash with service define values

https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-models/src/main/resources/models/discovery-2015-11-01-model.json#L425

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java [hunk 4]
```diff
@@ -708,6 +929,24 @@ public String toString() {
...
+        Builder listOfEnums(Collection<String> listOfEnums);
+
+        /**
+         * Sets the value of the ListOfEnums property for this object.
+         *
+         * @param listOfEnums
+         *        The new value for the ListOfEnums property for this object.
+         * @return Returns a reference to this object so that method calls can be chained together.
+         */
+        Builder listOfEnums(String... listOfEnums);
```
Comment (shorea) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java]: Why can't you set values via the enum?

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/EnumTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                .isEmpty();
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("Foo", "")))
+                .isEmpty();
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("EnumValue1", "Foo")))
+                .containsExactly(new SimpleImmutableEntry<>(EnumType.ENUM_VALUE1, EnumType.UNKNOWN));
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("EnumValue1", "")))
+                .containsExactly(new SimpleImmutableEntry<>(EnumType.ENUM_VALUE1, EnumType.UNKNOWN));
+    }
+
+    private EnumType simulateUnmarshallingEnum(String value) {
```
Comment (shorea) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/EnumTest.java]: Seems kind of a weird name. You're really just testing the builder behavior here.

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java [hunk 6]
```diff
@@ -125,21 +129,42 @@ public String getVariableName(String name) {
 
     @Override
     public String getEnumValueName(String enumValue) {
-        StringBuilder builder = new StringBuilder();
+        String result = enumValue;
 
-        String sanitizedEnumValue = enumValue.replace("::", ":").replace("/", "").replace("(", "")
-                                             .replace(")", "");
+        // Special cases
+        result = result.replaceAll("IoT", "IOT")
```
Comment (shorea) [inline codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: Should we treat these as customizations rather then put them in Core?

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java [hunk 7]
```diff
@@ -125,21 +129,42 @@ public String getVariableName(String name) {
...
+        // Remove extra underscores - multiple consecutive ones or those and the beginning/end of words
+        result = result.replaceAll("_+", "_") // Foo__Bar -> Foo_Bar
+                       .replaceAll("^_*([^_].*[^_])_*$", "$1"); // _Foo_ -> Foo
+
+        // Convert all lower-case words
+        result = StringUtils.upperCase(result);
+
+        if (!result.matches("^[A-Z][A-Z0-9_]*$")) {
+            String attempt = result;
+            log.warn(() -> "Invalid enum member generated for input '" + enumValue + "'. Best attempt: '" + attempt + "' If this "
```
Comment (shorea) [inline codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: Why not kill it here?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java [hunk 8]
```diff
@@ -163,16 +169,90 @@ private MethodSpec getValueForField() {
...
 
-    private CodeBlock getterStatement(MemberModel model) {
+    private Stream<MethodSpec> memberGetters(MemberModel member) {
+        List<MethodSpec> result = new ArrayList<>();
+
+        if (shouldGenerateEnumGetter(member)) {
+            result.add(enumMemberGetter(member));
+        }
+
+        result.add(nonEnumMemberGetter(member));
```
Comment (shorea) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java]: Name is kind of unintuitive as you'd call it for enums too it would just emit the string one.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java (hunk 3 above)
Comment (millems) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java]: +1

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java (hunk 4 above)
Comment (millems) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java]: We'll need to add support for that. We just don't have it yet, and I didn't want to make this PR any longer.

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/EnumTest.java (hunk 5 above)
Comment (millems) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/EnumTest.java]: I'm doing what the unmarshaller does. I'll add a comment.

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java (hunk 6 above)
Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: Do we have the feature in place to customize enum names? I didn't see anything like that.

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java (hunk 7 above)
Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: We have customizations that remove these values in a later part of code generation.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java (hunk 8 above)
Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java]: I'm trying to say "generate the non-enum getter for this member". For enums, that's a string. What name would you like me to use instead?

Comment (shorea) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java]: I'd just call it memberGetter since it's the primary getter for any type of field, the enum one is the convenience one.

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java (hunk 6 above)
Comment (shorea) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: I'm not sure, but we can add one. In general we try not to put service specific customizations in the code gen.

Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: That's fair - it's not good to have service-specific stuff in the generator. 

IoT in this case isn't service-specific. I'm not aware of any service using it in their enums, but they may in the future and we'd probably not notice it (generating IO_T instead).

It's just textORcsv that is service-specific in this PR. Because it's just this one example, I'd vote we follow the rule-of-three and add a customization when we have more cases in which we would use it. 

I don't think it is likely that more cases like this are going to happen. AWS service owners are warned now if their enums don't follow SCREAMING_CASE conventions. If something does happen to get through review: (a) we'll catch it before it's released and have the service owner fix their model, or (b) we won't catch it before it's released and at that point we can't change it anyway.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java (hunk 8 above)
Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AwsServiceModel.java]: :+1:

File: codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java (hunk 6 above)
Comment (shorea) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: Should the IoT case just be baked into the logic then. I.E. is there any value where we'd want to separate a single capital letter via underscore? If not can we just treat that as part of the previous word boundary?

Comment (millems) [reply codegen/src/main/java/software/amazon/awssdk/codegen/naming/DefaultNamingStrategy.java]: Sounds good to me.
